README.md
# websok

A high level websocket library. Use it when working with websockets
to avoid a few boilerplate and make your code look cleaner.

Supports IO and HTML websockets.

## Getting Started

```dart
/// For HTML:
/// import 'package:websok/html.dart';
///
/// For IO (Flutter, Dart, etc.)
import 'package:websok/io.dart';

/// Testing library.
import 'package:test/test.dart';

/// The received string.
String received;
/// Callback to execute when the function is over.
void onData(dynamic message) => received = message;

void main() {
  test('Performs a test websocket connection', () async {
    /// For HTML: IOWebsok -> HTMLWebsok
    final sok = IOWebsok(host: 'echo.websocket.org', tls: true)
      ..connect()
      ..listen(onData: onData);
    // Assets the connection.
    expect(sok.isActive, true);
    // Send a message.
    final message = 'Hello, world!';
    sok.send(message);
    // Check the message.
    await Future.delayed(Duration(seconds: 1), () => expect(received, message));
    // Close the connection after 1 sec.
    sok.close();
    // Assets the connection.
    expect(sok.isActive, false);
  });
}
```


CHANGELOG.md
## [2.1.1] - Fixes to HTML websocket.

* Fixed HTML websocket path.

## [2.1.0] - Add additional paths to the connection URL.

* Add additional paths to the connection URL.
* Isolated and created a `.url()` function that returns the connection URL.

## [2.0.5] - Copyright to main lib.

* Copyright to main lib.

## [2.0.4] - Flutter format

* Flutter format.

## [2.0.3] - Small fixes

* Naming issues.

## [2.0.2] - Small fixes

* Fixes to `README.md` and `websok_test.dart`
* Added example folder.
* Tried swaping some lib config to make it available on all platforms.

## [2.0.1] - Additional info to pubspec.yaml

* Added more info to pubspec.yaml

## [2.0.0] - Available on Web

* Websok is now also available on web.
* Use html.dart for HTML websockets, otherwise use io.dart.

## [1.1.0] - Available on Dart

* Websok is now also available on dart.
* Fixed TLS issue when no port provided.

## [1.0.0] - Initial version

* Initial release.


test/websok_test.dart
/// For HTML:
/// import 'package:websok/html.dart';
///
/// For IO (Flutter, Dart, etc.)
import 'package:websok/io.dart';

/// Testing library.
import 'package:test/test.dart';

/// The received string.
String received;
/// Callback to execute when the function is over.
void onData(dynamic message) => received = message;

void main() {
  test('Performs a test websocket connection', () async {
    /// For HTML: IOWebsok -> HTMLWebsok
    final sok = IOWebsok(host: 'echo.websocket.org', tls: true)
      ..connect()
      ..listen(onData: onData);
    // Assets the connection.
    expect(sok.isActive, true);
    // Assets the URL.
    expect(sok.url(), 'wss://echo.websocket.org:443/?');
    // Send a message.
    final message = 'Hello, world!';
    sok.send(message);
    // Check the message.
    await Future.delayed(Duration(seconds: 1), () => expect(received, message));
    // Close the connection after 1 sec.
    sok.close();
    // Assets the connection.
    expect(sok.isActive, false);
  });
}


lib/io.dart
/// Copyright (c) 2019 - Èrik Campobadal Forés
library websok.io;

import 'package:websok/websok.dart';
import 'package:web_socket_channel/io.dart';

/// For the use of @required.
import 'package:meta/meta.dart';

class IOWebsok extends Websok<IOWebSocketChannel> {
  /// HTTP request headers when [connect()] is called.
  final Map<String, String> headers;

  /// Determines the ping interval for the websocket.
  /// Null means no ping will be sent to the server.
  final Duration pingInterval;

  /// Creates a new IO Websocket.
  IOWebsok({
    @required String host,
    int port = -1,
    String path = '',
    Map<String, String> query = const <String, String>{},
    Iterable<String> protocols = const <String>[],
    this.headers = const <String, String>{},
    this.pingInterval,
    bool tls = false,
  }) : super(
          host: host,
          port: port,
          path: path,
          query: query,
          protocols: protocols,
          tls: tls,
        );

  /// Connects to the websocket given the following options.
  IOWebSocketChannel connectWith(String url, Iterable<String> protocols) =>
      IOWebSocketChannel.connect(
        url,
        protocols: protocols,
        headers: this.headers,
        pingInterval: this.pingInterval,
      );
}


lib/html.dart
/// Copyright (c) 2019 - Èrik Campobadal Forés
library websok.html;

import 'package:websok/websok.dart';
import 'package:web_socket_channel/html.dart';

/// For the use of @required.
import 'package:meta/meta.dart';

class HTMLWebsok extends Websok<HtmlWebSocketChannel> {
  /// controls what type is used for binary messages received by this socket.
  /// It defaults to [BinaryType.list], which causes binary messages to be delivered
  /// as [Uint8List]s. If it's [BinaryType.blob], they're delivered as [Blob]s instead.
  final BinaryType binaryType;

  /// Creates a new HTML web socket.
  HTMLWebsok({
    @required String host,
    int port = -1,
    String path = '',
    Map<String, String> query = const <String, String>{},
    Iterable<String> protocols = const <String>[],
    this.binaryType = BinaryType.list,
    bool tls = false,
  }) : super(
          host: host,
          port: port,
          path: path,
          query: query,
          protocols: protocols,
          tls: tls,
        );

  /// Connects to the websocket given the following options. Keep in mind not all
  /// of them will be used, it depends on [C] (HtmlWebSocketChannel).
  HtmlWebSocketChannel connectWith(String url, Iterable<String> protocols) =>
      HtmlWebSocketChannel.connect(
        url,
        protocols: protocols,
        binaryType: this.binaryType,
      );
}


lib/websock.dart
/// Copyright (c) 2019 - Èrik Campobadal Forés
library websok;

/// For the use of @required.
import 'package:meta/meta.dart';
/// Import the underlying package to use websockets.
// import 'package:web_socket_channel/io.dart';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:web_socket_channel/status.dart' as status;

/// The [Websock] class is used to create a new websocket connection.
/// It maintains all the state, socket connection and streams and
/// it abstracts it to a high level usage.
abstract class Websok<C extends WebSocketChannel>
{
  /// Stores the host of the websocket server.
  final String host;

  /// Stores the port of the websocket server.
  final int port;

  /// Stores if the connection is performed under TLS.
  final bool tls;

  /// Additional query string parameters attached to the URL.
  final Map<String, String> query;

  /// Determines the protocols used for the connection.
  final Iterable<String> protocols;

  /// Connects to the websocket given the following options.
  C connectWith(String url, Iterable<String> protocols);

  /// Determines if the connection is established.
  bool isActive = false;

  /// Stores the channel used in to send / receive messages.
  C channel;

  /// Creates a new websok instance and connects to the websocket immidiatly.
  /// If no port is provided, 80 wil be used in case tls = false, otherwise,
  /// 443 will be used.
  Websok({
    @required this.host,
    this.port = -1,
    this.query = const <String, String>{},
    this.protocols = const <String>[],
    this.tls = false
  });

  /// Connects to the websocket server.
  void connect() {
    final protocol = this.tls ? 'wss' : 'ws';
    final port = this.port != -1 ? this.port : (this.tls ? 443 : 80);
    var url = '$protocol://${this.host}:$port?';
    this.query.forEach((key, value) => url += '$key=$value&');
    this.channel = this.connectWith(url, this.protocols);
    this.isActive = true;
  }

  /// Listens for different events and executes their callback.
  void listen({ void onData(dynamic message), void onError(), void onDone(), bool cancelOnError }) =>
    this.channel.stream.listen(
      onData, onError: onError, onDone: onDone, cancelOnError: cancelOnError
    );

  /// Sends a message to the websocket.
  void send(String message) {
    if (this.isActive) this.channel.sink.add(message);
  }

  /// Closes the connection with the server.
  /// Returns a future which is completed when the stream sink has shut down.
  /// If cleaning up can fail, the error may be reported in the returned future,
  /// otherwise it completes with null.
  Future<dynamic> close([int code = status.goingAway, String reason]) {
    this.isActive = false;
    return this.channel.sink.close(code, reason);
  }
}


lib/websok.dart
/// Copyright (c) 2019 - Èrik Campobadal Forés
library websok;

/// For the use of @required.
import 'package:meta/meta.dart';

/// Import the underlying package to use websockets.
// import 'package:web_socket_channel/io.dart';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:web_socket_channel/status.dart' as status;

/// The [Websock] class is used to create a new websocket connection.
/// It maintains all the state, socket connection and streams and
/// it abstracts it to a high level usage.
abstract class Websok<C extends WebSocketChannel> {
  /// Stores the host of the websocket server.
  final String host;

  /// Stores the port of the websocket server.
  final int port;

  /// Stores if the connection is performed under TLS.
  final bool tls;

  /// Additional query string parameters attached to the URL.
  final Map<String, String> query;

  /// Determines the protocols used for the connection.
  final Iterable<String> protocols;

  /// Stores the URL path.
  final String path;

  /// Connects to the websocket given the following options.
  C connectWith(String url, Iterable<String> protocols);

  /// Determines if the connection is established.
  bool isActive = false;

  /// Stores the channel used in to send / receive messages.
  C channel;

  /// Creates a new websok instance and connects to the websocket immidiatly.
  /// If no port is provided, 80 wil be used in case tls = false, otherwise,
  /// 443 will be used.
  Websok({
    @required this.host,
    this.port = -1,
    this.path = '',
    this.query = const <String, String>{},
    this.protocols = const <String>[],
    this.tls = false,
  });

  /// Returns the URL used to connect tpo the websocket.
  String url() {
    final protocol = this.tls ? 'wss' : 'ws';
    final port = this.port != -1 ? this.port : (this.tls ? 443 : 80);
    final path = this.path.startsWith('/') ? this.path.substring(1) : this.path;
    var url = '$protocol://${this.host}:$port/$path?';
    this.query.forEach((key, value) => url += '$key=$value&');
    return url;
  }

  /// Connects to the websocket server.
  void connect() {
    this.channel = this.connectWith(this.url(), this.protocols);
    this.isActive = true;
  }

  /// Listens for different events and executes their callback.
  void listen({
    void onData(dynamic message),
    void onError(),
    void onDone(),
    bool cancelOnError,
  }) =>
      this.channel.stream.listen(
            onData,
            onError: onError,
            onDone: onDone,
            cancelOnError: cancelOnError,
          );

  /// Sends a message to the websocket.
  void send(String message) {
    if (this.isActive) this.channel.sink.add(message);
  }

  /// Closes the connection with the server.
  /// Returns a future which is completed when the stream sink has shut down.
  /// If cleaning up can fail, the error may be reported in the returned future,
  /// otherwise it completes with null.
  Future<dynamic> close([int code = status.goingAway, String reason]) {
    this.isActive = false;
    return this.channel.sink.close(code, reason);
  }
}
